### PR TITLE
Feature security2nd approach

### PR DESCRIPTION
**Pull request by Kondrjak**

### CODE REVIEW

File: src/main/java/neuefische/capstone/backend/rest/config/CustomizedRestMvcConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+
+@Component
+public class CustomizedRestMvcConfiguration implements RepositoryRestConfigurer {
+
+    @Override
+    public void configureRepositoryRestConfiguration(RepositoryRestConfiguration config, CorsRegistry cors) {
+        config.setBasePath("/api");
+        config.setRepositoryDetectionStrategy(
+                RepositoryDetectionStrategy.RepositoryDetectionStrategies.ANNOTATED
```
Comment (Kondrjak) [inline src/main/java/neuefische/capstone/backend/rest/config/CustomizedRestMvcConfiguration.java:17]: This prevents spring data rest from overwriting custom controllers

File: src/main/java/neuefische/capstone/backend/rest/ressource/CpGroupRes.java [hunk 2]
```diff
@@ -1,15 +1,14 @@
...
 import org.springframework.data.repository.PagingAndSortingRepository;
 import org.springframework.data.repository.query.Param;
 import org.springframework.data.rest.core.annotation.RepositoryRestResource;
 
+import java.util.List;
+
 @RepositoryRestResource(collectionResourceRel = "codepoint-groups", path = "codepoint-groups")
 public interface CpGroupRes extends PagingAndSortingRepository<CodepointGroup, Long> {
-
     List<CodepointGroup> findByName(@Param("name") String name);
```
Comment (belluma) [inline src/main/java/neuefische/capstone/backend/rest/ressource/CpGroupRes.java:12]: what is this @Param doing?

File: src/main/java/neuefische/capstone/backend/security/service/CredentialService.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Autowired
+    public CredentialService(CredentialRepo repo) {
+        this.repo = repo;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String username)
+            throws UsernameNotFoundException {
+        Optional<Credential> maybeCredential = repo.findByUsername(username);
```
Comment (belluma) [inline src/main/java/neuefische/capstone/backend/security/service/CredentialService.java:27]:
```suggestion
      
```

File: src/main/java/neuefische/capstone/backend/security/service/CredentialService.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        return maybeCredential
```
Comment (belluma) [inline src/main/java/neuefische/capstone/backend/security/service/CredentialService.java:28]:
```suggestion
        return repo.findByUsername(username)
```
no need for the other line, orElseThrow takes care of that

File: src/test/java/neuefische/capstone/backend/security/controler/LoginControllerThrowsTest.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Test
+    void login_WithValidCredentials_ShouldReturnValidJwt(){
+        // GIVEN
+        String username = "user";
+        String password = "correct";
+        String hashedPw = new BCryptPasswordEncoder().encode(password);
+        Credential credentialToSave = Credential.builder().username(username).password(hashedPw).build();
+        credentialRepo.save(credentialToSave);
+        Credential nowValidCredential = Credential.builder().username(username).password(password).build();
```
Comment (belluma) [inline src/test/java/neuefische/capstone/backend/security/controler/LoginControllerThrowsTest.java:60]: if you use @With annotation on your model you could shorten this line to 
``` Credential nowValidCredential = credentialToSave.withPassword(password);```

File: src/test/java/neuefische/capstone/backend/security/service/CredentialServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    private final CredentialRepo credentialRepo = mock(CredentialRepo.class);
+    private final CredentialService credentialService = new CredentialService(credentialRepo);
+
+    @Test
+    void loadUserByUsername_shouldThrow_whenUserIsNotInRepo() {
+        // GIVEN
+        when(credentialRepo.findById("invalid")).thenReturn(Optional.empty());
+
+        // WHEN
+        // credentialService.loadUserByUsername("invalid")
```
Comment (belluma) [inline src/test/java/neuefische/capstone/backend/security/service/CredentialServiceTest.java:24]:
```suggestion
       
```

File: src/test/java/neuefische/capstone/backend/testUtils/TestJwtHeaders.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    private CredentialRepo credentialRepo;
+
+    /**
+     * In order to go through the jwt token authentication process,
+     * all test http mappings need an authenticated users header
+     *
+     * @return authenticated users http headers for controller tests
+     */
+    public HttpHeaders get(int port){
+        Optional<Credential> maybeCredential = credentialRepo.findByUsername(username);
```
Comment (belluma) [inline src/test/java/neuefische/capstone/backend/testUtils/TestJwtHeaders.java:36]:
```suggestion
return  credentialRepo.findByUsername(username).map().orElse();
```

**Review by belluma — APPROVED** [review]
well done

File: src/main/java/neuefische/capstone/backend/DevQuizApplication.java [hunk 8]
```diff
@@ -1,13 +1,13 @@
...
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 
 @SpringBootApplication
-public class DevCapstoneBackendApplication {
+public class DevQuizApplication {
 
     public static void main(String[] args) {
-        SpringApplication.run(DevCapstoneBackendApplication.class, args);
+        SpringApplication.run(DevQuizApplication.class, args);
     }
```
Comment (Cobas91) [inline src/main/java/neuefische/capstone/backend/DevQuizApplication.java:12]: Rename it?

**Review by Cobas91 — APPROVED** [review]
Add .DS_Store to your .gitignore file. Mac is creating these files automaticly in every folder.

File: src/main/java/neuefische/capstone/backend/rest/ressource/CpGroupRes.java (hunk 2 above)
Comment (Kondrjak) [reply src/main/java/neuefische/capstone/backend/rest/ressource/CpGroupRes.java:12]: telling Spring Data Rest to bind the method variable name to querry param name in the created controller

File: src/main/java/neuefische/capstone/backend/DevQuizApplication.java (hunk 8 above)
Comment (Kondrjak) [reply src/main/java/neuefische/capstone/backend/DevQuizApplication.java:12]: done ^^

**Pull request merged**